.vsts.release.yml

```diff
@@ -27,61 +35,117 @@ parameters:
   default: true
   displayName: Test Proxy Agent
 
-variables:
-  ${{ if parameters.buildStageOnly }}:
-    releaseBranch: ''
-  ${{ else }}:
-    releaseBranch: releases/${{ parameters.version }}
-
 extends:
   template: .azure-pipelines/pipeline.yml
   parameters:
-    branch:  ${{ variables.releaseBranch }}
+    branch:  $[ stageDependencies.Verify_release.Set_variables.outputs['SetReleaseVariables.releaseBranch'] ]
     componentDetection: false
     test: ${{ not(parameters.skipTests) }}
     sign: true
     publishArtifacts: true
     testProxyAgent: ${{ parameters.testProxyAgent }}
+    stageDependencies:
+      - Verify_release
+      - Create_Release_Branch
+    stageCondition: |
+      and(
+        succeeded('Verify_release'),
+        in(dependencies.Create_Release_Branch.result, 'Succeeded', 'Skipped')
+      )
 
     preBuildStages:
-    - ${{ if and(not(parameters.buildStageOnly), eq(variables['Build.SourceBranch'], 'refs/heads/master')) }}:
-      - stage: Create_Release_Branch
+    - stage: Verify_release
+      displayName: Verify it's a release run
+      pool:
+        vmImage: ubuntu-latest
+      jobs:
+      - job: Set_variables
+        displayName: Set release-specific variables
+        steps:
+        - pwsh: |
+            $isBuildStageOnly = [System.Convert]::ToBoolean('${{ parameters.buildStageOnly }}')
+            $buildReason = '$(Build.Reason)'
+
+            $currentSprint = (Invoke-WebRequest https://whatsprintis.it -Headers @{"Accept" = "application/json" } | ConvertFrom-Json)
+            $isReleaseWeek = $currentSprint.week -eq 3
+            Write-Host "isReleaseWeek = $isReleaseWeek"
+
+            $isRelease = ($buildReason -eq 'Manual' -and !$isBuildStageOnly) -or ($buildReason -eq 'Schedule' -and !$isReleaseWeek)
+            Write-Host "isRelease = $isRelease"
+            Write-Host "##vso[task.setVariable variable=isRelease;isOutput=true]$isRelease"
+
+            $isScheduledRelease = $isRelease -and $buildReason -eq 'Schedule'
+            Write-Host "isScheduledRelease = $isScheduledRelease"
+
+            if ($isRelease) {
+              if ($isScheduledRelease) {
+                $agentVersion = "3.$($currentSprint.sprint).0"
+              } else {
+                $agentVersion = "${{ parameters.version }}"
```

Comment:
What if agent version is not specified in parameters?

Reply:
Will be used default version: https://github.com/microsoft/azure-pipelines-agent/pull/4480/files#diff-22b8a6edaa7fd4e7111cefba1c8e0b432b9eeeeddc90cde7177d739e9b3b2226R12

But since this param was required before, we probably should fail it wasn't set...

Reply:
Added validation at runtime